federation-js/src/joinSpec.ts

```diff
@@ -0,0 +1,146 @@
+  // Build a map of original service name to generated name
+  const sanitizedServiceNames: Record<string, string> = Object.create(null);
+
+  function uniquifyAndSanitizeGraphQLName(name: string) {
+    // Transforms to ensure valid graphql `Name`
+    const alphaNumericUnderscoreOnly = name.replace(/[^_a-zA-Z0-9]/g, '_');
+    const noNumericFirstChar = alphaNumericUnderscoreOnly.match(/^[0-9]/)
+      ? '_' + alphaNumericUnderscoreOnly
+      : alphaNumericUnderscoreOnly;
+    const noUnderscoreNumericEnding = noNumericFirstChar.match(/_[0-9]+$/)
+      ? noNumericFirstChar + '_'
+      : noNumericFirstChar;
+
+    // toUpper not really necessary but follows convention of enum values
+    const toUpper = noUnderscoreNumericEnding.toLocaleUpperCase();
+
+    // Uniquifying post-transform
+    const nameCount = nameMap.get(toUpper);
+    if (nameCount) {
+      // Collision - bump counter by one
+      nameMap.set(toUpper, nameCount + 1);
+      const uniquified = `${toUpper}_${nameCount + 1}`;
+      // We also now need another entry for the name we just generated
+      nameMap.set(uniquified, 1);
+      sanitizedServiceNames[name] = uniquified;
+      return uniquified;
+    } else {
+      nameMap.set(toUpper, 1);
+      sanitizedServiceNames[name] = toUpper;
+      return toUpper;
+    }
+  }
+
+  return {
+    sanitizedServiceNames,
+    JoinGraphEnum: new GraphQLEnumType({
+      name: 'join__Graph',
+      values: Object.fromEntries(
+        serviceList.map((service) => [
+          uniquifyAndSanitizeGraphQLName(service.name),
+          { value: service },
+        ]),
+      ),
+    }),
+  };
+}
+
+function getJoinFieldDirective(JoinGraphEnum: GraphQLEnumType) {
+  return new GraphQLDirective({
+    name: 'join__field',
+    locations: [DirectiveLocation.FIELD_DEFINITION],
+    args: {
+      graph: {
+        type: JoinGraphEnum,
+      },
+      requires: {
+        type: FieldSetScalar,
+      },
+      provides: {
+        type: FieldSetScalar,
+      },
+    },
+  });
+}
+
+function getJoinOwnerDirective(JoinGraphEnum: GraphQLEnumType) {
+  return new GraphQLDirective({
+    name: 'join__owner',
+    locations: [DirectiveLocation.OBJECT, DirectiveLocation.INTERFACE],
+    args: {
+      graph: {
+        type: new GraphQLNonNull(JoinGraphEnum),
+      },
+    },
+  });
+}
+
+export function getJoins(serviceList: ServiceDefinition[]) {
+  const { sanitizedServiceNames, JoinGraphEnum } = getJoinGraphEnum(serviceList);
+  const JoinFieldDirective = getJoinFieldDirective(JoinGraphEnum);
+  const JoinOwnerDirective = getJoinOwnerDirective(JoinGraphEnum);
+
+  const JoinTypeDirective = new GraphQLDirective({
+    name: 'join__type',
+    locations: [DirectiveLocation.OBJECT, DirectiveLocation.INTERFACE],
+    isRepeatable: true,
+    args: {
+      graph: {
+        type: new GraphQLNonNull(JoinGraphEnum),
+      },
+      key: {
+        type: FieldSetScalar,
+      },
+    },
+  });
+
+  return {
+    sanitizedServiceNames,
```

Comment:
How about choosing a name for this field (and other instances of this token) that makes it clear which direction the map goes, and that the sanitization is specifically for the enum values? (And avoids the "service" terminology.) Like `graphNameToEnumValueName`?